### DISCUSSION TITLE
Error: text/template.Execute: cannot convert non-concrete value string

### DISCUSSION THREAD

**Discussion by SimonEbner** [cue-lang/cue#1122] [Q&A]
Hi everybody. I'm hitting an error with template.Execute that I'm not able to resolve: `text/template.Execute: cannot convert non-concrete value string.`
Disclaimer: I'm very new to cue, so I may doing it completely wrong.

Reproducible code can be found here:
https://github.com/SimonEbner/cuelang-template-error

**Details**:
I have a main file `test.cue`. It uses the transform pattern in a somewhat convoluted way:
```
import "example.com/foo"
_arg: {
	in: {
		param1: 123
		param2: "x86_64"
		param3: "xyz"
	}
}
_transform: {
	_in: [...]
	_intermediate: [ for x in _in {x & _arg}]
	out: [ for x in _intermediate {x.result}]
}
_transform & {_in: [foo.a]}
```

Inside the `foo` package, I use the template as such:
```
package foo
import "text/template"
#Input: {
	param2: string
	param3: string
	...
}
a: {
	in: #Input
	result:  template.Execute("test", in)
}
```

The error it produces is the following
```
_in.0.in.param2: error in call to text/template.Execute: cannot convert non-concrete value string:
    ./foo/foo.cue:15:11
    ./foo/foo.cue:8:2
_in.0.in.param3: error in call to text/template.Execute: cannot convert non-concrete value string:
    ./foo/foo.cue:15:11
    ./foo/foo.cue:9:2
```

However, I'm able to fix it in two very unintuitive ways:

**1. Variation: Reducing definition**
Replace `#Input` inside `foo/foo.cue` with the reduced definition `#Input{param2: string, ...}` (removing param3).
Running `cue eval test.cue -c` finishes cleanly with
```
out: ["test"]
```

**2. Remove template.Execute**
Replace result: `template.Execute("test", in)` in `foo/foo.cue` with simply `result: in`.
Running `cue eval test.cue -c` finishes cleanly with
```
out: [{
    param1: 123
    param2: "x86_64"
    param3: "xyz"
}]
```
The part that puzzles me is that the two changes don't seem to be changing anything fundamentally, but make it compile cleanly with `-c`.
Can somebody help me?

**Comment by verdverm** [comment]
the repo doesn't look to be using `text/template` like you describe.

What happens if you change `a: <...>` into a definition `#a: <...>`?

**Reply by SimonEbner** [reply]
My bad, I had actually submited variation 2 from above. I restored the error condition and turned `a: ` into `#a`. I pushed it to the repository. The error is the same

**Comment by myitcv** [comment]
Hi @SimonEbner - welcome to the CUE project!

As a quick FYI, for multi-file repros like yours above, it can be easier to use the `txtar` format (saves creating a repro, or copy-pasting multiple files into an issue). See [the wiki](https://github.com/cue-lang/cue/wiki/Creating-test-or-performance-reproducers) for details.

Looking at your example above, you can somewhat simplify your "functions" using embedded scalars. See https://github.com/cue-lang/cue/discussions/484#discussioncomment-959924 for an example. Note that we are considering some syntactic sugar to make this even cleaner on the eye.

Looking at your specific problem it appears to be an instance of https://github.com/cue-lang/cue/issues/1095

**Reply by SimonEbner** [reply]
Thanks a lot for all inputs.
